src/lib/storage/lz4/lz4_encoder.hpp

```diff
@@ -0,0 +1,126 @@
+#include "utils/enum_constant.hpp"
+
+#include "lib/lz4hc.h"
+
+namespace opossum {
+
+class LZ4Encoder : public SegmentEncoder<LZ4Encoder> {
+ public:
+  static constexpr auto _encoding_type = enum_c<EncodingType, EncodingType::LZ4>;
+  static constexpr auto _uses_vector_compression = false;
+
+  template <typename T>
+  std::shared_ptr<BaseEncodedSegment> _on_encode(const std::shared_ptr<const ValueSegment<T>>& value_segment) {
+    const auto alloc = value_segment->values().get_allocator();
+    const auto num_elements = value_segment->size();
+    DebugAssert(num_elements <= std::numeric_limits<int>::max(),
+                "Trying to compress a ValueSegment with more "
+                "elements than fit into an int.");
+
+    // TODO(anyone): when value segments switch to using pmr_vectors, the data can be copied directly instead of
+    // copying it element by element
+    auto values = pmr_vector<T>{alloc};
+    values.reserve(num_elements);
+    auto null_values = pmr_vector<bool>{alloc};
+    null_values.reserve(num_elements);
+
+    // copy values and null flags from value segment
+    auto iterable = ValueSegmentIterable<T>{*value_segment};
+    iterable.with_iterators([&](auto it, auto end) {
+      for (; it != end; ++it) {
+        auto segment_value = *it;
+        values.emplace_back(segment_value.value());
+        null_values.emplace_back(segment_value.is_null());
+      }
+    });
+
+    // LZ4 compression
+    const auto input_size = static_cast<int>(values.size() * sizeof(T));
+    auto output_size = LZ4_compressBound(input_size);
+    auto compressed_data = pmr_vector<char>{alloc};
+    compressed_data.resize(static_cast<size_t>(output_size));
+    const int compression_result = LZ4_compress_HC(reinterpret_cast<char*>(values.data()), compressed_data.data(),
+                                                   input_size, output_size, LZ4HC_CLEVEL_MAX);
+    Assert(compression_result > 0, "LZ4 compression failed");
+
+    auto data_ptr = std::allocate_shared<pmr_vector<char>>(alloc, std::move(compressed_data));
+    auto null_values_ptr = std::allocate_shared<pmr_vector<bool>>(alloc, std::move(null_values));
+
+    return std::allocate_shared<LZ4Segment<T>>(alloc, data_ptr, null_values_ptr, nullptr, compression_result,
+                                               input_size, num_elements);
+  }
+
+  std::shared_ptr<BaseEncodedSegment> _on_encode(
+      const std::shared_ptr<const ValueSegment<std::string>>& value_segment) {
+    const auto alloc = value_segment->values().get_allocator();
+    const auto num_elements = value_segment->size();
+    DebugAssert(num_elements <= std::numeric_limits<int>::max(),
+                "Trying to compress a ValueSegment with more "
+                "elements than fit into an int.");
+
+    // copy values and null flags from value segment
+    auto values = pmr_vector<char>{alloc};
+    auto null_values = pmr_vector<bool>{alloc};
+    null_values.reserve(num_elements);
```

Comment:
1.) Please also reserve (or resize according to @Felerius's point) `values`. Something sane as `4*num_elements` should already have an good impact.
2.) Please ensure that this encoding as well as decoding is tested with vectors of empty strings. This is very common in real-world data and might yield problems with your handling.

Reply:
I don't think the reserve will make much of a difference. Take a look at this [example benchmark](http://quick-bench.com/aplSNPOV_-7cKL6sP_gNhFqBtKg).

Reply:
I think this benchmark is fairer: http://quick-bench.com/CRpktHw2mMVZL94P7v_7n6ixpGA

And I'd argue it's worth the added single line.

Reply:
We summed the total number of characters since we also use that number to handle the edge case of only empty strings.

Reply:
But why not collecting the number during construction?